docs/src/main/asciidoc/extension-maturity-model.adoc

```diff
@@ -0,0 +1,93 @@
+
+[id="extension-maturity-model"]
+= A maturity model for Quarkus extensions
+include::_attributes.adoc[]
+:diataxis-type: concept
+:categories: writing-extensions
+:topics: extensions
+:summary: Quarkus extensions can do a lot, or a little. This guide explains some of the capabilities extension authors may wish to include.
+////
+The document header ends at the first blank line. Do not remove the blank line between the header and the abstract summary.
+////
+
+What makes a good Quarkus extension? What capabilities is a Quarkus extension expected to provide?
+
+* It actually works
+** Works in JVM mode
+** Works in dev mode
+** Works in native
+* Dev service (if there is an external service dependency)
+* Developer joy
+** Configuration support
+** Use build-time application knowledge to remove boilerplate
+* Dev UI
+* Supersonic subatomic performance
+** Use build-time application knowledge to eliminate wasteful runtime code paths
+* Codestart application template
+
+The order in this model isn't exact. Different developers will have different views on what capabilities are most important. You may wish to (say) prioritise performance over enhancing your extension's Dev UI tile. That's fine!
```

Comment:
the list is good - but afaics the only thing that "fits" a maturity model (as in something that is progressively more) is "works in jvm" and "works in native" (too). The rest a checkbox features.

i.e. does it help to call it a extension maturity model or does it confuse more ?

Reply:
I wondered the same thing, because you're right that the progression is not a strict one. But I think it does make sense to talk about the maturity of an extension. In this context it's a synonym for completeness, but completeness wouldn't be helpful term to use, because it makes it sound like people aren't 'allowed' to do extensions unless they do everything. And 'checklist' has a similar problem - it doesn't have *enough* progression. 

Another way of thinking about it is that we're trying to distinguish a really good extension from an average one, but I don't think any words like 'quality' would be helpful, because there's too much judgement. This should be an aspiration list, not a shame list.

Reply:
yeah, I don't have too strong opinion ....just that in past we kept getting stuck trying to "rank" it .

Alternative suggestion to maturity model I can think of is Capability Model: Emphasizes the capabilities or features that are completed or available.

another is "Implementation Levels: Describes the levels at which different parts of the project have been implemented." but that also "ranks"

Reply:
The good thing about "maturity model" is that it will be really obvious to people what it means. "This page tells me what my extension will be like when it starts, and where it could eventually end up if I continue working on it." "Capability model" or "Implementation levels" would need some words around it to try to explain to people what the purpose of the page is, because they're not familiar concepts in the same way. 

The bad thing about maturity model is that this isn't a maturity model in the strictest sense, because the progression/dependencies on lower levels isn't strict enough. And there definitely could be some debates about the relative position of a few things. 

We could change the page title to something generic like extension-capability-model but leave the in-page title as "maturity model". That way if we decide to change the the title later on we don't have a problem with the url?

Reply:
+1 on using capability model for the URL.

I've looked at the staircase multiple times and it keeps rubbing me the wrong way that the two cornerstone capabilities Quarkus pushes/priorities is the last two steps before last.

It would be more coherent and signal of focus if we move devservice/devui higher up maturity scale as we really want programming model and performance to be early priority.

With that its:

Works In JVM -> Works in devmode -> works in native mode -> configuration -> joyful programming model -> supersonic performance -> provide a dev service -> dev UI -> codestart

Yes, we could debate additional ordering but with that at least config/joyful/perf is better balanced with respect to what implicit message the stair case inherently brings.

Reply:
Ordering will be very different from one extension to another. We should not crystalize on that. 

The only ordering we have is: works in JVM -> works in native (as the opposite is rarely true)

I would use "capabilities", or "traits" as I mentioned in another comment. 

In another comment, Max shows some radar graph. While tricky to set it right, I think it is slightly better:

- no ordering 
- extensions can decide to focus on different aspects (some dev only extension do not care or performance or cloud readiness, or even native (mailpit for example))